File: src/main/java/io/github/intellij/dlanguage/features/DBraceMatcher.java
package io.github.intellij.dlanguage.features;

import com.intellij.lang.BracePair;
import com.intellij.lang.PairedBraceMatcher;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.tree.TokenSet;
import io.github.intellij.dlanguage.psi.DTokenSets;
import io.github.intellij.dlanguage.psi.DlangTypes;
import io.github.intellij.dlanguage.psi.DlangTypes;
import io.github.intellij.dlanguage.psi.DTokenSets;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class DBraceMatcher implements PairedBraceMatcher {
    private static final BracePair[] PAIRS = new BracePair[]{
            new BracePair(DlangTypes.PARENTHESES_LEFT, DlangTypes.PARENTHESES_RIGHT, false),
            new BracePair(DlangTypes.BRACES_LEFT, DlangTypes.BRACES_RIGHT, true),
            new BracePair(DlangTypes.BRACKETS_LEFT, DlangTypes.BRACKETS_RIGHT, false),
    };

    private static final TokenSet alwaysMatch = TokenSet.create(DlangTypes.OP_BRACES_LEFT);

    @Override
    public BracePair[] getPairs() {
        return PAIRS;
    }

    @Override
    public boolean isPairedBracesAllowedBeforeType(@NotNull final IElementType lbraceType,
                                                   @Nullable final IElementType contextType) {
        return alwaysMatch.contains(lbraceType)
            || DTokenSets.WHITESPACES.contains(contextType)
            || DTokenSets.LINE_COMMENTS.contains(contextType)
            || DTokenSets.BLOCK_COMMENTS.contains(contextType)
            || DlangTypes.OP_PAR_RIGHT == contextType
            || null == contextType;
    }

    @Override
    public int getCodeConstructStart(final PsiFile file, final int openingBraceOffset) {
        return openingBraceOffset;
    }
}

File: src/main/kotlin/io/github/intellij/dlanguage/features/DBraceMatcher.kt
package io.github.intellij.dlanguage.features

import com.intellij.lang.BracePair
import com.intellij.lang.PairedBraceMatcher
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IElementType
import com.intellij.psi.tree.TokenSet
import io.github.intellij.dlanguage.psi.DTokenSets
import io.github.intellij.dlanguage.psi.DlangTypes

class DBraceMatcher : PairedBraceMatcher {
    override fun getPairs(): Array<BracePair> {
        return PAIRS
    }

    override fun isPairedBracesAllowedBeforeType(
        lbraceType: IElementType,
        contextType: IElementType?
    ): Boolean {
        return (alwaysMatch.contains(lbraceType)
                || DTokenSets.WHITESPACES.contains(contextType)
                || DTokenSets.LINE_COMMENTS.contains(contextType)
                || DTokenSets.BLOCK_COMMENTS.contains(contextType)) || DlangTypes.OP_PAR_RIGHT === contextType || null == contextType
    }

    override fun getCodeConstructStart(file: PsiFile, openingBraceOffset: Int): Int {
        return openingBraceOffset
    }

    companion object {
        private val PAIRS = arrayOf(
            BracePair(DlangTypes.OP_PAR_LEFT, DlangTypes.OP_PAR_RIGHT, false),
            BracePair(DlangTypes.OP_BRACES_LEFT, DlangTypes.OP_BRACES_RIGHT, true),
            BracePair(DlangTypes.OP_BRACKET_LEFT, DlangTypes.OP_BRACKET_RIGHT, false)
        )
        private val alwaysMatch = TokenSet.create(DlangTypes.OP_BRACES_LEFT)
    }
}
